### DISCUSSION TITLE
Discussion: reference equality operator

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#540] [Language Ideas] [answered]
dotnet/roslyn#8409 suggests to use `===` and `!==` operators as the reference equality operator. Although, since `is null` covered half of the use cases, is it possible to use the same operator to do a reference equality check with a generic expression?
```cs
var equal = e1 is e2;
````

**Comment by DavidArno** [comment]
I think your `is` syntax:
```cs
var equal = e1 is e2;
```
is a huge improvement on `===`, as I suspect anyone who's been near JavaScript, reads the latter as "really equal" or similar.

**Comment by orthoxerox** [comment]
So `e1 is e2` is literally `ReferenceEquals(e1, e2)`? Are there that many cases when you must test your objects for reference equality, but their `==` operator is overridden to mean something different?

**Comment by alrz** [comment]
As it turns out `ReferenceEquals` call, already [generates optimized code](https://tryroslyn.azurewebsites.net/#f:>ilr/K4Zwlgdg5gBAygTxAFwKYFsDcBYAUAB2ACMAbMAYxnJIEMQQYBhGAbzxg5kNIpiIHt+JGAFkAFPyIArVOWQx+MADQLps+QA8AlK3ad9YAGZiASqkOoATqgjlUAUQCOwGiRASl2rS2vJgliBhkS2BUTABfX39Aw1cQML0OcLxwoA=). Closing.

**Comment by jnm2** [comment]
Please reopen? The main benefit I was looking forward to was replacing all my `ReferenceEquals(x, y)` with `x is y` syntax!

**Comment by DavidArno** [comment]
Seconding @jnm2: I thought the whole point to this proposal was for better syntax for `object.ReferenceEquals(x, y)`. I'd expect the bytecode to be identical, so don't see this as a performance issue.

**Comment by alrz** [comment]
The first issue that arises here is that it silently changes the semantics of `is` based on RHS being a pattern (or active pattern) or not.

**Comment by dstarkowski** [comment]
Won't this conflict with current `is` operator?

    String String = "a";
    String Other = "a";
    var equal = Other is String;

**Comment by alrz** [comment]
No binding changes are suggested as part of this proposal. You could already use `is (String)` to force it to be resolved as an expression.

**Comment by YaakovDavis** [comment, marked as answer]
> The main benefit I was looking forward to was replacing all my `ReferenceEquals(x, y)` with `x is y` syntax

I would disagree that `x is y` is beneficiary over `ReferenceEquals(x, y)`. The latter is explicit, while the former is vague. With all pattern matching going around, one will have to remember what `is` means in this case.

Related to that, there were multiple proposals of allowing the user to reinterpret `is`. This proposal conflicts with those.

I'd rather have flexibility over `is` reinterpretation, than a hardcoded "reference equality" semantics.

**Comment by ronnygunawan** [comment]
@alrz `x is y` syntax no longer possible because RHS must be a constant value. On top of that, we now have records which overrides equality operators by default.

Please reopen https://github.com/dotnet/roslyn/issues/8409?

**Reply by alrz** [reply]
Feel free to open a discussion here in dotnet/csharplang. 

I do believe it still has merits but I guess `===` brings back memories from JS so the majority dislike it.
